contracts/mirror/src/contract.rs

```diff
@@ -0,0 +1,484 @@
+    _data: String,
+) -> ContractResult<Response<NeutronMsg>> {
+    let id = get_id_from_request_memo(&req)?;
+    deps.api.debug(
+        format!(
+            "WASMDEBUG: sudo_error: ack received: {:?} id: {:?}",
+            req, id
+        )
+        .as_str(),
+    );
+    let mut bond = BONDS.load(deps.storage, id)?;
+    bond.state = BondState::Bonded;
+    BONDS.save(deps.storage, id, &bond)?;
+    let attrs = vec![
+        attr("action", "sudo_error"),
+        attr("id", id.to_string()),
+        attr("state", bond.state.to_string()),
+    ];
+    Ok(response("sudo_error", CONTRACT_NAME, attrs))
+}
+
+fn sudo_timeout(
+    deps: DepsMut<NeutronQuery>,
+    req: RequestPacket,
+) -> ContractResult<Response<NeutronMsg>> {
+    let id = get_id_from_request_memo(&req)?;
+    deps.api.debug(
+        format!(
+            "WASMDEBUG: sudo_timeout: ack received: {:?} id: {:?}",
+            req, id
+        )
+        .as_str(),
+    );
+    let mut bond = BONDS.load(deps.storage, id)?;
+    bond.state = BondState::Bonded;
+    BONDS.save(deps.storage, id, &bond)?;
+    let attrs = vec![
+        attr("action", "sudo_timeout"),
+        attr("id", id.to_string()),
+        attr("state", bond.state.to_string()),
+    ];
+    Ok(response("sudo_timeout", CONTRACT_NAME, attrs))
+}
+
+fn sudo_response(
+    deps: DepsMut<NeutronQuery>,
+    req: RequestPacket,
+    data: Binary,
+) -> ContractResult<Response<NeutronMsg>> {
+    let request_data: FungibleTokenPacketData = from_json(req.data.clone().unwrap())?; // must present as there is only IBC transfer for this contract
+    let id: u64 = get_id_from_request_memo(&req)?;
+    deps.api.debug(
+        format!(
+            "WASMDEBUG: sudo_response: sudo received: {:?} {:?} {:?} id: {:?}",
+            req, data, request_data, id
+        )
+        .as_str(),
+    );
+    BONDS.remove(deps.storage, id);
+    let attrs = vec![
+        attr("action", "sudo_response"),
+        attr("request", format!("{:?}", req)),
+        attr("id", id.to_string()),
+    ];
+    Ok(response("sudo_response", CONTRACT_NAME, attrs))
+}
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn migrate(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    _msg: MigrateMsg,
+) -> ContractResult<Response<NeutronMsg>> {
+    let version: semver::Version = CONTRACT_VERSION.parse()?;
+    let storage_version: semver::Version =
+        cw2::get_contract_version(deps.storage)?.version.parse()?;
+    if storage_version < version {
+        cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    }
+    Ok(Response::new())
+}
+
+fn get_id_from_request_memo(req: &RequestPacket) -> ContractResult<u64> {
+    let request_data: FungibleTokenPacketData = from_json(req.data.clone().unwrap())?; // must present as there is only IBC transfer for this contract
+    let id: u64 = request_data
+        .memo
+        .parse()
+        .map_err(|_| ContractError::InvalidMemo)?;
```

Comment:
Do we really need this error for invalid memo, if we always serialize this memo ourselves with a call to `to_string()`? I think it is safe to unwrap here, too :)

Reply:
the serialization of the memo is outside the contract so I'd better left the error here

Reply:
Don't we fill memo in `contracts/mirror/src/contract.rs:308`?